### ISSUE TITLE
Option to failFast on zip and sequence

### ISSUE THREAD

**Issue by scabug** [scala/bug#8994]
It should be possible to supply a flag to fail fast on both zip and sequence operations.

val a = Future{Thread.sleep(5000);1}
val b = Future{Thread.sleep(1000); throw new Exception("test")}
val c = a zip b

c will fail in 5 seconds with the exception from B.

Most clients would probably prefer it to fail in 1 second even if that means that they might get an exception from Future b instead of a possible exception from Future a.

**Comment by julienrf** [comment]
I think this is an important issue. In addition to this, I’ve noticed that `zip` is not commutative “fail-wise”.

See this snippet:

~~~ scala
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{ Await, Future, blocking }
import scala.concurrent.duration.Duration

def failing() = Future.failed(new Exception("KO"))
def succeeding() = Future(blocking(Thread.sleep(5000)))

System.currentTimeMillis() / 1000 // : Long = 1603205724
Await.ready(failing().zip(succeeding()), Duration.Inf)
System.currentTimeMillis() / 1000 // : Long = 1603205724
Await.ready(succeeding().zip(failing()), Duration.Inf)
System.currentTimeMillis() / 1000 // : Long = 1603205729
~~~

Note that executing `failing().zip(succeeding())` completes instantaneously, whereas executing `succeeding().zip(failing())` completes only after `succeeding()` completes.

I think we should fix `zip` such that `a.zip(b)` fails as soon as `a` or `b` fails. We should also make sure that `sequence` and `traverse` are fixed.

**Comment by julienrf** [comment]
@johnynek do you want to contribute your fix?

**Comment by SethTisue** [comment]
perhaps @viktorklang has a take on this

**Comment by viktorklang** [comment]
@SethTisue @julienrf I definitely don't mind fail-fast behavior added—assuming that it can be done in a resource-efficient way which does not lead to performance degradations for existing code.

**Comment by johnynek** [comment]
In the code linked above, we use futures to wait on hadoop jobs finishing. In such cases, the real cost is running a job you know you don't need. This is not the usual case with futures in scala.

My solution is fairly straight forward. I think you can improve the performance but my assumption is the linked code above will reduce performance of zip.

**Comment by julienrf** [comment]
> In the code linked above, we use futures to wait on hadoop jobs finishing. In such cases, the real cost is running a job you know you don't need. This is not the usual case with futures in scala.

I think this is out of the scope of `Future` (although I would love to see this concern addressed in the scala-library!)

> My solution is fairly straight forward. I think you can improve the performance but my assumption is the linked code above will reduce performance of zip.

OK, then we need to run some benchmarks to see if the performance reduction would be significant or not.

**Comment by viktorklang** [comment]
Note that we're talking about changing the implementation of `zipWith`—since `zip` uses `zipWith`. Also, I do not see a way to implement it generically on `Future` since it will require a concrete implementation of `Promise`, which `trait Future` cannot assume.

**Comment by eloots** [comment]
Give the users of the library a choice; those who want 'the performance' can choose the current implementation. For other use cases that require a fail fast behaviour, use a new implementation (`zipWithFailFast`, `Future.sequenceFailFast`, ...).
There are valid use cases for both.

I noticed `Future.sequence` not failing fast a long time ago. See here for an implementation of [`Future.sequenceFailFast`](https://github.com/scalacenter/effective-scala-slide-code/pull/3/files#diff-7af96373389cfcb8dffd68ae4664b3b1d4776f03a3df65fb01f8f98df7559701) (Scala 3 code).

**Comment by julienrf** [comment]
Closed by scala/scala#9655
